numpy/core/src/multiarray/buffer.c

```diff
@@ -248,14 +260,17 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
             /* Insert padding manually */
             if (*offset > new_offset) {
                 PyErr_SetString(PyExc_RuntimeError,
-                                "This should never happen: Invalid offset in "
-                                "buffer format string generation. Please "
-                                "report a bug to the Numpy developers.");
+                                "The buffer interface does not support "
+                                "overlapping fields");
```

Comment:
Doesn't this path also trigger for fields that are not in sequence?

Reply:
For future reference, one more datapoint: The HDF5 datatype specification also stores "compound datatypes" as an ordered list of `(name, format, offset)` tuples. So that is now 3 cases where structured datatypes are stored that way: numpy dtypes, MPI serialization, and HDF5 files. 

I want to record these examples here in case in the future we want to argue to the CPython devs that we should add a new PEP3118 specification style for structured types, that is equivalent to a list of tuples.